File: Python/백준/Gold/1005. ACM Craft/ACM Craft.py
from collections import deque
import sys
input = sys.stdin.readline

T = int(input())
for _ in range(T):
    N, K = map(int, input().split())
    D = list(map(int, input().split()))
    adj = [[] for _ in range(N)]
    indegree = [0]*N
    for _ in range(K):
        X, Y = map(int, input().split())
        adj[X-1].append(Y-1)
        indegree[Y-1] += 1
    W = int(input()) - 1
    dp = D[:]
    q = deque([i for i in range(N) if indegree[i]==0])
    while q:
        now = q.popleft()
        for nxt in adj[now]:
            dp[nxt] = max(dp[nxt], dp[now]+D[nxt])
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                q.append(nxt)
    print(dp[W])

File: Java/백준/Gold/1005. ACM Craft/ACM Craft.java
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            int N = sc.nextInt();
            int K = sc.nextInt();
            int[] time = new int[N+1];
            int[] dp = new int[N+1];
            ArrayList<Integer>[] adj = new ArrayList[N+1];
            int[] indegree = new int[N+1];
            for (int i=1; i<=N; i++) {
                time[i] = sc.nextInt();
                dp[i] = time[i];
                adj[i] = new ArrayList<>();
            }
            for (int i=0; i<K; i++) {
                int X = sc.nextInt();
                int Y = sc.nextInt();
                adj[X].add(Y);
                indegree[Y]++;
            }
            int W = sc.nextInt();
            Queue<Integer> q = new LinkedList<>();
            for (int i=1; i<=N; i++) if (indegree[i]==0) q.add(i);
            while (!q.isEmpty()) {
                int now = q.poll();
                for (int nxt : adj[now]) {
                    dp[nxt] = Math.max(dp[nxt], dp[now]+time[nxt]);
                    indegree[nxt]--;
                    if (indegree[nxt]==0) q.add(nxt);
                }
            }
            System.out.println(dp[W]);
        }
    }
}
